services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt

```diff
@@ -32,6 +58,264 @@ class HakemusService(
+    private fun assertGeometryValidity(
+        areas: List<ApplicationArea>?,
+        customMessageOnFailure: (GeometriatDao.InvalidDetail) -> String
+    ) {
+        if (areas != null) {
+            geometriatDao.validateGeometriat(areas.map { it.geometry })?.let {
+                throw ApplicationGeometryException(customMessageOnFailure(it))
+            }
+        }
+    }
+
+    /**
+     * Assert that the customers match and that the contacts in the update request are hanke users
+     * of the application hanke.
+     */
+    private fun assertYhteystiedotValidity(
+        applicationEntity: ApplicationEntity,
+        updateRequest: HakemusUpdateRequest
+    ) {
+        val customersWithContacts = updateRequest.customersByRole()
+        ApplicationContactType.entries.forEach {
+            assertYhteystietoValidity(
+                applicationEntity.id!!,
+                it,
+                applicationEntity.yhteystiedot[it],
+                customersWithContacts[it]
+            )
+        }
+
+        assertYhteyshenkilotValidity(
+            applicationEntity.hanke,
+            customersWithContacts.values
+                .filterNotNull()
+                .flatMap { it.contacts.map { contact -> contact.hankekayttajaId } }
+                .toSet()
+        ) {
+            "Invalid hanke user/users received when updating application id=${applicationEntity.id}, invalid hankeKayttajaIds=$it"
+        }
+    }
+
+    /**
+     * If the yhteystieto is present in the application and in the request, the ids must match. If
+     * the yhteystieto is not present in the application, it means that the request adds a new
+     * yhteystieto, and it cannot have an id. If the yhteystieto is not present in the request, it
+     * means that the application yhteystieto is removed.
+     */
+    private fun assertYhteystietoValidity(
+        applicationId: Long,
+        rooli: ApplicationContactType,
+        hakemusyhteystietoEntity: HakemusyhteystietoEntity?,
+        customerWithContacts: CustomerWithContactsRequest?
+    ) {
+        if (
+            hakemusyhteystietoEntity != null &&
+                customerWithContacts != null &&
+                customerWithContacts.customer.yhteystietoId != hakemusyhteystietoEntity.id
+        ) {
+            // ids don't match
+            throw InvalidHakemusyhteystietoException(
+                applicationId,
+                rooli,
+                hakemusyhteystietoEntity.id,
+                customerWithContacts.customer.yhteystietoId
+            )
+        }
+        if (
+            hakemusyhteystietoEntity == null &&
+                customerWithContacts?.customer?.yhteystietoId != null
+        ) {
+            // new yhteystieto has an id when it shouldn't
+            throw InvalidHakemusyhteystietoException(
+                applicationId,
+                rooli,
+                null,
+                customerWithContacts.customer.yhteystietoId
+            )
+        }
```

Comment:
I think this would be simpler if the positive cases were handled first. Then all others can throw an exception. The positive cases are:
1. CustomerWithContacts is missing.
2. The customer in customerWithContacts doesn't have an ID.
3. The IDs are the same.
```suggestion
        if (customerWithContacts == null || customerWithContacts.customer.yhteystietoId == null) {
            return
        }
        if (customerWithContacts.customer.yhteystietoId == hakemusyhteystietoEntity?.id) {
            return
        }
        throw InvalidHakemusyhteystietoException(
            applicationId,
            rooli,
            hakemusyhteystietoEntity?.id,
            customerWithContacts.customer.yhteystietoId
        )
```

Reply:
I refactored this as you suggested but combined the first two ifs into one.